File: Solutions/Python/number_of_islands.py
"""
    [MEDIUM]
    200. Number of Islands

    Concepts:
    - matrix
    - breadth-first search

    Stats:
        Runtime | 275 ms    [Beats 24.79%]
        Memory  | 26.49 MB  [Beats 15.15%]
"""

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        """
            Iterate over every cell

            When land found, run BFS
            on cell until water reached

            When finished, increment number of islands
        """
        m, n = len(grid), len(grid[0])
        num_islands = 0
        explored = set()

        def bfs(y: int, x: int) -> set:
            nonlocal explored

            queue = deque([(y, x)])
            visited = set()

            while queue:
                len_queue = len(queue)

                for _ in range(len_queue):
                    i, j = queue.popleft()

                    for di, dj in [(0, 1), (1, 0), (0, -1), (-1, 0)]:
                        if (0 <= (ni := i + di) < m and
                            0 <= (nj := j + dj) < n and
                            grid[ni][nj] == "1" and
                            (ni, nj) not in visited):

                            queue.append((ni, nj))
                            visited.add((ni, nj))

            return visited

        for i in range(m):
            for j in range(n):
                if grid[i][j] == "1" and (i, j) not in explored:
                    explored.update(bfs(i, j))
                    num_islands += 1

        return num_islands

File: Solutions/Cpp/number_of_islands.cpp
/*
    [MEDIUM]
    200. Number of Islands

    Concepts:
    - matrix
    - breadth-first search

    Stats:
        Runtime | 39 ms     [Beats 17.03%]
        Memory  | 29.13 MB  [Beats 11.83%]
*/

class Solution {
public:
    int n, m;
    const std::array<const std::array<const int, 2>, 4> moves = {
        std::array<const int, 2>{0, -1},
        std::array<const int, 2>{0, 1},
        std::array<const int, 2>{-1, 0},
        std::array<const int, 2>{1, 0}
    };

    inline int coords_to_idx(int i, int j) {
        return (i * n) + (j % n);
    }
    
    int numIslands(vector<vector<char>>& grid) {
        m = grid.size();
        n = grid[0].size();
        std::unordered_set<int> visited = {};
        int num_islands = 0;

        std::function<void(int, int)> bfs;

        bfs = [&](int row, int col) {
            std::queue<std::array<int, 2>> q;
            q.push({row, col});
            int len_q, ni, nj, idx;

            while (!q.empty()) {
                len_q = q.size();

                for (int _ = 0; _ < len_q; _++) {
                    auto [i, j] = q.front();
                    q.pop();

                    for (auto const [di, dj] : moves) {
                        ni = i + di, nj = j + dj;
                        idx = coords_to_idx(ni, nj);
                        if (
                            0 <= ni && ni < m &&
                            0 <= nj && nj < n &&
                            grid[ni][nj] == '1' &&
                            !visited.contains(idx)
                        ) {
                            q.push({ni, nj});
                            visited.insert(idx);
                        }
                    }
                }
            }
        };

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == '1' && !visited.contains(coords_to_idx(i, j))) {
                    bfs(i, j);
                    num_islands++;
                }
            }
        }

        return num_islands;
    }
};